changes/views.py

```diff
@@ -46,3 +52,147 @@ class ChangeGetViewSet(viewsets.ReadOnlyModelViewSet):
+        try:
+            data = utils.json_decode(request.body)
+        except ValueError as e:
+            return JsonResponse({'reason': "JSON decode error",
+                                'details': six.text_type(e)}, status=400)
+
+        try:
+            identity_id = data['identity']
+            optout_reason = data['optout_reason']
+            optout_source = data['optout_source']
+        except KeyError as e:
+            return JsonResponse({'reason': '"identity", "optout_type", '
+                                 '"optout_reason" and "optout_source" must be '
+                                 'specified.'
+                                 }, status=400)
+
+        registration = Registration.objects.get(mother_id=identity_id)
+        if registration.data.get('msg_receiver'):
+            fire_optout_receiver_type_metric(registration.data['msg_receiver'])
+
+        fire_optout_reason_metric(optout_reason)
+
+        if registration.data.get('msg_type'):
+            fire_optout_message_type_metric(registration.data['msg_type'])
+
+        fire_optout_source_metric(optout_source)
+
+        return JsonResponse({})
+
+
+def fire_optout_reason_metric(reason):
+    from registrations.tasks import fire_metric
+
+    if reason not in settings.OPTOUT_REASONS:
+        reason = 'other'
+
+    fire_metric.apply_async(kwargs={
+        "metric_name": 'optout.reason.%s.sum' % reason,
+        "metric_value": 1.0
+    })
+
+    def search_optouts_reason():
+        result = utils.search_optouts({"reason": reason})
+        return len(list(result))
+
+    total_key = 'optout.reason.%s.total.last' % reason
+    total = get_or_incr_cache(
+        total_key,
+        search_optouts_reason)
+    fire_metric.apply_async(kwargs={
+        'metric_name': total_key,
+        'metric_value': total,
+    })
+
+
+def fire_optout_source_metric(source):
+    from registrations.tasks import fire_metric
+
+    # remove the _public part
+    source_short = source.split('_')[0]
+
+    fire_metric.apply_async(kwargs={
+        "metric_name": 'optout.source.%s.sum' % source_short,
+        "metric_value": 1.0
+    })
+
+    def search_optouts_source():
+        result = utils.search_optouts({"request_source": source})
+        return len(list(result))
+
+    total_key = 'optout.source.%s.total.last' % source_short
+    total = get_or_incr_cache(
+        total_key,
+        search_optouts_source)
+    fire_metric.apply_async(kwargs={
+        'metric_name': total_key,
+        'metric_value': total,
+    })
+
+
+def fire_optout_receiver_type_metric(msg_receiver):
+    from registrations.tasks import fire_metric
+
+    fire_metric.apply_async(kwargs={
+        "metric_name": 'optout.receiver_type.%s.sum' % msg_receiver,
+        "metric_value": 1.0
+    })
+
+    def search_optouts_receiver_type():
+        result = utils.search_optouts()
+
+        identities = []
+        for data in result:
+            identities.append(data['identity'])
```

Comment:
We can probably use list comprehension here, and use a set instead of a list to ensure that we don't have duplicates. Like `identities = set(data['identity'] for data in result)`